qcodes/tests/dataset/test_data_set_cache.py

```diff
@@ -331,6 +333,150 @@ def test_cache_complex_array_param_in_1d(experiment, DAC, channel_array_instrume
+        'dummy_channel_inst_ChanA_dummy_parameter_with_setpoints_complex': (n_points_outer, n_points_inner, pws_n_points)
+    }
+
+    if not cache_too_small:
+        assert meas._shapes == expected_shapes
+
+    with meas.run(write_in_background=bg_writing) as datasaver:
+        dataset = datasaver.dataset
+        _assert_parameter_data_is_identical(dataset.get_parameter_data(), dataset.cache.data())
+        n_points_measured = 0
+        for v1 in np.linspace(-1, 1, n_points_outer):
+            for v2 in np.linspace(-1, 1, n_points_inner):
+                n_points_measured += 1
+                DAC.ch1.set(v1)
+                DAC.ch2.set(v2)
+                meas_vals = [(param, param.get()) for param in meas_parameters]
+
+                datasaver.add_result((DAC.ch1, v1),
+                                     (DAC.ch2, v2),
+                                     *meas_vals)
+                datasaver.flush_data_to_database(block=True)
+                param_data_trees = dataset.get_parameter_data()
+                cache_data_trees = dataset.cache.data()
+
+                _assert_partial_cache_is_as_expected(
+                    cache_data_trees,
+                    expected_shapes,
+                    n_points_measured,
+                    param_data_trees,
+                    cache_too_small
+                )
+    cache_data_trees = dataset.cache.data()
+    param_data_trees = dataset.get_parameter_data()
+
+    _assert_completed_cache_is_as_expected(cache_data_trees,
+                                           param_data_trees,
+                                           flatten=cache_too_small)
+
+
+def _assert_completed_cache_is_as_expected(
+        cache_data_trees,
+        param_data_trees,
+        flatten=False):
+    for outer_key, cache_data_tree in cache_data_trees.items():
+        for inner_key, cache_data in cache_data_tree.items():
+            if flatten:
+                assert_array_equal(
+                    cache_data.flatten(),
+                    param_data_trees[outer_key][inner_key].flatten()
+                )
+            else:
+                assert_array_equal(
+                    cache_data,
+                    param_data_trees[outer_key][inner_key]
+                )
+
+
+def _assert_partial_cache_is_as_expected(
+        cache_data_trees,
+        expected_shapes,
+        n_points_measured,
+        param_data_trees,
+        cache_expected_too_small=False
+):
+    assert sorted(cache_data_trees.keys()) == sorted(expected_shapes.keys())
+    for outer_key, cache_data_tree in cache_data_trees.items():
+        exshape = expected_shapes[outer_key]
+        if len(exshape) > 2:
+            array_shape = np.prod(exshape[2:])
+        else:
+            array_shape = 1
+
+        for inner_key, cache_data in cache_data_tree.items():
+            if not cache_expected_too_small:
+                assert cache_data.shape == exshape
```

Comment:
do i understand it right?
```suggestion
            if not cache_expected_too_small:
                assert cache_data.shape == exshape
            else:
                assert cache_data.size > np.prod(exshape)  # sanity check of cache_expected_too_small flag usage
```

Reply:
Nope this is not always true, Only after a number of measurements have been performed

Reply:
ah, yes, indeed, sorry